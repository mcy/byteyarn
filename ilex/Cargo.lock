# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "annotate-snippets"
version = "0.9.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ccaf7e9dfbb6ab22c82e473cd1a8a7bd313c19a5b7e40970f3d89ef5a5c9e81e"
dependencies = [
 "unicode-width",
 "yansi-term",
]

[[package]]
name = "bitflags"
version = "2.4.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "327762f6e5a765692301e5bb513e0d9fef63be86bbc14528052b1cd3e6f03e07"

[[package]]
name = "byteyarn"
version = "0.2.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "a7534301c0ea17abb4db06d75efc7b4b0fa360fce8e175a4330d721c71c942ff"

[[package]]
name = "camino"
version = "1.1.6"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "c59e92b5a388f549b863a7bea62612c09f24c8393560709a54558a9abdfb3b9c"

[[package]]
name = "diff"
version = "0.1.13"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "56254986775e3233ffa9c4d7d3faaf6d36a2c09d30b20687e9f88bc8bafc16c8"

[[package]]
name = "ilex"
version = "0.1.0"
dependencies = [
 "annotate-snippets",
 "bitflags",
 "byteyarn",
 "camino",
 "pretty_assertions",
 "unicode-segmentation",
 "unicode-xid",
]

[[package]]
name = "pretty_assertions"
version = "1.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "af7cee1a6c8a5b9208b3cb1061f10c0cb689087b3d8ce85fb9d2dd7a29b6ba66"
dependencies = [
 "diff",
 "yansi",
]

[[package]]
name = "unicode-segmentation"
version = "1.10.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "1dd624098567895118886609431a7c3b8f516e41d30e0643f03d94592a147e36"

[[package]]
name = "unicode-width"
version = "0.1.11"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e51733f11c9c4f72aa0c160008246859e340b00807569a0da0e7a1079b27ba85"

[[package]]
name = "unicode-xid"
version = "0.2.4"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f962df74c8c05a667b5ee8bcf162993134c104e96440b663c8daa176dc772d8c"

[[package]]
name = "winapi"
version = "0.3.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5c839a674fcd7a98952e593242ea400abe93992746761e38641405d28b00f419"
dependencies = [
 "winapi-i686-pc-windows-gnu",
 "winapi-x86_64-pc-windows-gnu",
]

[[package]]
name = "winapi-i686-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ac3b87c63620426dd9b991e5ce0329eff545bccbbb34f3be09ff6fb6ab51b7b6"

[[package]]
name = "winapi-x86_64-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "712e227841d057c1ee1cd2fb22fa7e5a5461ae8e48fa2ca79ec42cfc1931183f"

[[package]]
name = "yansi"
version = "0.5.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "09041cd90cf85f7f8b2df60c646f853b7f535ce68f85244eb6731cf89fa498ec"

[[package]]
name = "yansi-term"
version = "0.1.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "fe5c30ade05e61656247b2e334a031dfd0cc466fadef865bdcdea8d537951bf1"
dependencies = [
 "winapi",
]
