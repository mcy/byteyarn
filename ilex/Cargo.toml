[package]
name = "ilex"
version = "0.4.1"
description = "quick and easy lexers for C-like languages"

edition.workspace = true
authors.workspace = true
homepage.workspace = true
repository.workspace = true
keywords.workspace = true
license.workspace = true

[dependencies]
byteyarn = { version = "0.4", path = "../byteyarn" }
twie = { version = "0.4", path = "../twie" }

annotate-snippets = "0.10.0"
camino = "1.1.6"
dashmap = "5.5.3"
num-traits = "0.2.17"
similar-asserts = "1.5.0"
rustc_apfloat = "0.2.0"  # By eddyb's recommendation.
unicode-xid = "0.2.4"
