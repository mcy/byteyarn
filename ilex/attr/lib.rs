//! Implementation detail of `ilex`.

use proc_macro::TokenStream;

// This helper exists only to make the #[spec] field attributes inert.
#[doc(hidden)]
#[proc_macro_derive(derive_hack, attributes(named, rule))]
pub fn derive(_: TokenStream) -> TokenStream {
  TokenStream::new()
}

proc2decl::bridge! {
  /// Generates a lexer spec struct.
  ///
  /// This macro generates the type of struct described in the
  /// [crate documentation][crate]. The syntax is as follows.
  ///
  /// ```ignore
  /// use ilex::rule::Keyword;
  /// use ilex::Lexeme;
  ///
  /// /// My cool spec.
  /// #[ilex::spec]
  /// struct MySpec {
  ///   #[named("...")]
  ///   #[rule(/* ... */)]
  ///   dollar: Lexeme<Keyword> = "$",
  /// }
  /// ```
  ///
  /// The type of each field must be a [`Lexeme`] with a [`Rule`] type as its
  /// parameter. There are two special attributes that can follow.
  ///
  /// - `#[named]` makes the rule into a *named* rule. This name can be used by
  ///   diagnostics, and corresponds to calling `Spec::named_rule()`.
  ///
  /// - `#[rule]` is the value to use to construct the rule, which must be
  ///   `Into<R>`, where `R` is the type inside `Lexeme` (so, above, the rule
  ///   value must be `Into<Keyword>`). By default, this value is the name of the
  ///   rule, to make the common case of declaring a keyword as simple as writing
  ///   `nullptr: Lexeme<Keyword>`, assuming Rust itself doesn't already use that
  ///   keyword.
  ///
  /// Note that *order matters* for the fields: when breaking a tie between two
  /// potential tokens of the same length, the first one in the struct will win.
  /// In practice, this means you should put keywords before identifiers.
  ///
  /// Additionally, the following functions will be defined for the `MySpec` type.
  ///
  /// ```
  /// # struct Spec;
  /// # struct MySpec;
  /// # fn norun(_: i32) {
  /// impl MySpec {
  ///   /// Gets the global instance of this spec.
  ///   pub fn get() -> &'static Self {
  ///     // ...
  /// #   todo!()
  ///   }
  ///
  ///   /// Gets the actual compiled spec.
  ///   pub fn spec(&self) -> &Spec {
  ///     // ...
  /// #   todo!()
  ///   }
  /// }
  /// # }
  /// ```
  ///
  // God cross-trait links suck.
  /// [`Lexeme`]: https://docs.rs/ilex/latest/ilex/struct.Lexeme.html
  /// [`Rule`]: https://docs.rs/ilex/latest/ilex/rule/trait.Rule.html
  /// [crate]: https://docs.rs/ilex
  macro #[spec] => ilex::__spec__;
}
