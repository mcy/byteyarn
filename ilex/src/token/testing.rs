//! Lexer testing helpers.
//!
//! This type provides testing-oriented matchers for matching on a
//! [`TokenStream`][`crate::token::Stream`].
//!
//! These matchers are intended for writing *tests*. To write a parser, you\
//! should use [`Cursor`][crate::token::Cursor] instead.

use std::fmt;
use std::fmt::DebugStruct;
use std::fmt::Display;
use std::ops::Range;

use std::format_args as f;

use byteyarn::Yarn;

use crate::file::Context;
use crate::file::Span;
use crate::file::Spanned;
use crate::lexer::rule;
use crate::lexer::Lexeme;
use crate::token;
use crate::token::Any;
use crate::token::Cursor;
use crate::token::Sign;

/// Matches a token cursor against a list of token matchers.
///
/// Returns `Ok(())` when they match, otherwise returns a pretty-printed error
/// showing what went wrong.
pub fn recognize_tokens<'a>(
  ctx: &Context,
  stream: Cursor,
  matchers: impl IntoIterator<Item = &'a Matcher>,
) -> Result<(), impl fmt::Debug> {
  let mut state = MatchState {
    ctx,
    errors: String::new(),
    stack: Vec::new(),
    error_count: 0,
  };

  struct DebugBy(String);
  impl fmt::Debug for DebugBy {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
      f.write_str(&self.0)
    }
  }

  zip_eq(&mut state, matchers, stream, |state, ours, theirs| {
    ours.recognizes(state, theirs, ctx)
  });
  state.finish().map_err(DebugBy)
}

/// A matcher for a chunk of text from the input source.
///
/// This is slightly more general than a span, since it can specify the content
/// of the text and the offsets separately, and optionally. `Text` values are
/// intended to *recognize* various spans.
///
/// `&str` and `Range<usize>` are both convertible to `Text`.
pub struct Text {
  text: Option<Yarn>,
  range: Option<Range<usize>>,
}

impl Text {
  /// Returns a matcher that recognizes all spans.
  pub fn any() -> Self {
    Text {
      text: None,
      range: None,
    }
  }

  /// Returns a matcher that recognizes spans with the given text.
  pub fn new(text: impl Into<Yarn>) -> Self {
    Text {
      text: Some(text.into()),
      range: None,
    }
  }

  /// Returns a matcher that recognizes spans with the given byte range.
  pub fn range(range: Range<usize>) -> Self {
    Text {
      text: None,
      range: Some(range),
    }
  }

  /// Returns a matcher that recognizes spans with the given byte range and
  /// text.
  pub fn text_and_range(text: impl Into<Yarn>, range: Range<usize>) -> Self {
    Text {
      text: Some(text.into()),
      range: Some(range),
    }
  }

  /// Returns whether this span recognizes a particular span.
  fn recognizes(&self, span: Span, ctx: &Context) -> bool {
    !self
      .text
      .as_ref()
      .is_some_and(|text| text != span.text(ctx))
      && !self
        .range
        .as_ref()
        .is_some_and(|range| range != &span.range(ctx).unwrap())
  }
}

impl<Y: Into<Yarn>> From<Y> for Text {
  fn from(value: Y) -> Self {
    Text::new(value.into())
  }
}

impl fmt::Debug for Text {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    match (&self.text, &self.range) {
      (Some(text), Some(range)) => write!(f, "{text:?} @ {range:?}"),
      (Some(text), None) => fmt::Debug::fmt(text, f),
      (None, Some(range)) => write!(f, "<any> @ {range:?}"),
      (None, None) => f.write_str("<any>"),
    }
  }
}

pub struct Matcher {
  which: Option<Lexeme<rule::Any>>,
  span: Text,
  comments: Vec<Text>,
  kind: Kind,
}

enum Kind {
  Eof,
  Unexpected,
  Keyword,
  Ident {
    name: Text,
    prefix: Option<Text>,
    suffix: Option<Text>,
  },
  Quoted {
    content: Vec<token::Content<Text>>,
    delims: (Text, Text),
    prefix: Option<Text>,
    suffix: Option<Text>,
  },
  Digital {
    mant: DigitalMatcher,
    exps: Vec<DigitalMatcher>,
    suffix: Option<Text>,
  },
  Delimited {
    delims: (Text, Text),
    tokens: Vec<Matcher>,
  },
}

#[derive(Debug)]
struct DigitalMatcher {
  radix: u8,
  sign: Option<(Sign, Text)>,
  digits: Vec<Text>,
  prefix: Option<Text>,
}

impl Lexeme<rule::Keyword> {
  pub fn matcher(self, text: impl Into<Text>) -> Matcher {
    Matcher {
      which: Some(self.any()),
      span: text.into(),
      comments: Vec::new(),
      kind: Kind::Keyword,
    }
  }
}

impl Lexeme<rule::Ident> {
  pub fn matcher(self, name: impl Into<Text>) -> Matcher {
    Matcher {
      which: Some(self.any()),
      span: Text::any(),
      comments: Vec::new(),
      kind: Kind::Ident {
        name: name.into(),
        prefix: None,
        suffix: None,
      },
    }
  }
}

impl<Y: Into<Yarn>> From<Y> for token::Content<Text> {
  fn from(value: Y) -> Self {
    Self::Lit(value.into().into())
  }
}

impl Lexeme<rule::Quoted> {
  pub fn matcher<C: Into<token::Content<Text>>>(
    self,
    quotes: (impl Into<Text>, impl Into<Text>),
    content: impl IntoIterator<Item = C>,
  ) -> Matcher {
    Matcher {
      which: Some(self.any()),
      span: Text::any(),
      comments: Vec::new(),
      kind: Kind::Quoted {
        content: content.into_iter().map(C::into).collect(),
        delims: (quotes.0.into(), quotes.1.into()),
        prefix: None,
        suffix: None,
      },
    }
  }
}

impl Lexeme<rule::Digital> {
  pub fn matcher<B, T>(self, radix: u8, blocks: B) -> Matcher
  where
    B: IntoIterator<Item = T>,
    T: Into<Text>,
  {
    Matcher {
      which: Some(self.any()),
      span: Text::any(),
      comments: Vec::new(),
      kind: Kind::Digital {
        mant: DigitalMatcher {
          radix,
          sign: None,
          digits: blocks.into_iter().map(Into::into).collect(),
          prefix: None,
        },
        exps: Vec::new(),
        suffix: None,
      },
    }
  }
}

impl Lexeme<rule::Bracket> {
  pub fn matcher(
    self,
    brackets: (impl Into<Text>, impl Into<Text>),
    contents: Vec<Matcher>,
  ) -> Matcher {
    Matcher {
      which: Some(self.any()),
      span: Text::any(),
      comments: Vec::new(),
      kind: Kind::Delimited {
        tokens: contents,
        delims: (brackets.0.into(), brackets.1.into()),
      },
    }
  }
}

impl Matcher {
  pub fn eof() -> Self {
    Matcher {
      which: Some(Lexeme::eof().any()),
      span: Text::any(),
      comments: Vec::new(),
      kind: Kind::Eof,
    }
  }

  pub fn unexpected() -> Self {
    Matcher {
      which: None,
      span: Text::any(),
      comments: Vec::new(),
      kind: Kind::Unexpected,
    }
  }

  pub fn span(mut self, text: impl Into<Text>) -> Self {
    self.span = text.into();
    self
  }

  pub fn prefix(mut self, text: impl Into<Text>) -> Self {
    match &mut self.kind {
      Kind::Ident { prefix, .. } | Kind::Quoted { prefix, .. } => {
        *prefix = Some(text.into());
      }
      Kind::Digital { mant, .. } => mant.prefix = Some(text.into()),
      _ => unreachable!(),
    }

    self
  }

  pub fn suffix(mut self, text: impl Into<Text>) -> Self {
    match &mut self.kind {
      Kind::Ident { suffix, .. }
      | Kind::Quoted { suffix, .. }
      | Kind::Digital { suffix, .. } => {
        *suffix = Some(text.into());
      }
      _ => unreachable!(),
    }

    self
  }

  pub fn sign(mut self, value: Sign, text: impl Into<Text>) -> Self {
    match &mut self.kind {
      Kind::Digital { mant, .. } => mant.sign = Some((value, text.into())),
      _ => unreachable!(),
    }

    self
  }

  pub fn exponent<T: Into<Text>>(
    mut self,
    radix: u8,
    delim: impl Into<Text>,
    sign: Option<(Sign, Text)>,
    blocks: impl IntoIterator<Item = T>,
  ) -> Self {
    match &mut self.kind {
      Kind::Digital { exps, .. } => exps.push(DigitalMatcher {
        radix,
        sign,
        digits: blocks.into_iter().map(Into::into).collect(),
        prefix: Some(delim.into()),
      }),
      _ => unreachable!(),
    }

    self
  }

  pub fn comments<I>(mut self, iter: I) -> Self
  where
    I: IntoIterator,
    I::Item: Into<Text>,
  {
    self.comments.extend(iter.into_iter().map(Into::into));
    self
  }

  fn recognizes(&self, state: &mut MatchState, tok: token::Any, ctx: &Context) {
    state.match_spans("token span", &self.span, tok.span(ctx));

    zip_eq(state, &self.comments, tok.comments(ctx), |state, t, s| {
      state.match_spans("comment", t, s);
    });

    match (&self.kind, tok) {
      (Kind::Eof, Any::Eof(..))
      | (Kind::Unexpected, Any::Unexpected(..))
      | (Kind::Keyword, Any::Keyword(..)) => {}
      (
        Kind::Ident {
          name,
          prefix,
          suffix,
        },
        Any::Ident(tok),
      ) => {
        state.match_spans("identifier name", name, tok.name());
        state.match_options("prefix", prefix.as_ref(), tok.prefix());
        state.match_options("suffix", suffix.as_ref(), tok.suffix());
      }
      (
        Kind::Quoted {
          delims,
          content,
          prefix,
          suffix,
        },
        Any::Quoted(tok),
      ) => {
        let (open, close) = tok.delimiters();
        state.match_spans("open quote", &delims.0, open);
        state.match_spans("close quote", &delims.1, close);
        state.match_options("prefix", prefix.as_ref(), tok.prefix());
        state.match_options("suffix", suffix.as_ref(), tok.suffix());

        zip_eq(
          state,
          content,
          tok.raw_content(),
          |state, ours, theirs| match (ours, theirs) {
            (token::Content::Lit(t), token::Content::Lit(s)) => {
              state.match_spans("string content", t, s)
            }
            (token::Content::Esc(t, ours), token::Content::Esc(s, theirs)) => {
              state.match_spans("string escape", t, s);
              if ours != &theirs {
                state.error(f!(
                  "wrong decoded escape; want {:?}, got {:?}",
                  ours,
                  theirs
                ));
              }
            }
            _ => state.error("mismatched string content types"),
          },
        );
      }
      (Kind::Digital { mant, exps, suffix }, Any::Digital(tok)) => {
        let recognize = |state: &mut MatchState,
                         mch: &DigitalMatcher,
                         tok: token::Digital| {
          if mch.radix != tok.radix() {
            state.error(f!(
              "wrong radix; want {:?}, got {:?}",
              mch.radix,
              tok.radix()
            ));
          }
          state.match_any_options(
            "sign",
            mch.sign.as_ref().map(|(s, _)| s),
            tok.sign(),
            |&a, b| a == b,
          );
          state.match_options(
            "sign span",
            mch.sign.as_ref().map(|(_, sp)| sp),
            tok.sign_span(),
          );
          state.match_options("prefix", mch.prefix.as_ref(), tok.prefix());
          zip_eq(state, &mch.digits, tok.digit_blocks(), |state, t, s| {
            state.match_spans("digit block", t, s);
          });
        };

        recognize(state, mant, tok);
        zip_eq(state, exps, tok.exponents(), |state, t, s| {
          recognize(state, t, s);
        });

        state.match_options("suffix", suffix.as_ref(), tok.suffix());
      }
      (Kind::Delimited { delims, tokens }, Any::Bracket(tok)) => {
        state.match_spans("open delimiter", &delims.0, tok.open());
        state.match_spans("close delimiter", &delims.1, tok.close());

        zip_eq(state, tokens, tok.contents(), |state, ours, theirs| {
          ours.recognizes(state, theirs, ctx)
        });
      }
      _ => state.error("mismatched token types"),
    }
  }
}

impl fmt::Debug for Matcher {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    let print_spans = matches!(
      std::env::var("ILEX_SPANS").as_deref(),
      Ok("ranges" | "text")
    );

    let req_field = |d: &mut DebugStruct, name, span| {
      if print_spans {
        d.field(name, span);
      }
    };

    let opt_field = |d: &mut DebugStruct, name, span: &Option<Text>| {
      if print_spans && span.is_some() {
        d.field(name, span.as_ref().unwrap());
      }
    };

    let vec_field = |d: &mut DebugStruct, name, spans: &Vec<Text>| {
      if !spans.is_empty() {
        d.field(name, spans);
      }
    };

    let name = match &self.kind {
      Kind::Eof => "Eof",
      Kind::Unexpected => "Unexpected",
      Kind::Keyword => "Keyword",
      Kind::Ident { .. } => "Ident",
      Kind::Quoted { .. } => "Quoted",
      Kind::Digital { .. } => "Digital",
      Kind::Delimited { .. } => "Delimited",
    };

    let name = match self.which {
      Some(l) => format!("{name}({})", l.index()),
      None => name.into(),
    };

    let mut d = f.debug_struct(&name);

    match &self.kind {
      Kind::Ident {
        name,
        prefix,
        suffix,
      } => {
        req_field(&mut d, "name", name);
        opt_field(&mut d, "prefix", prefix);
        opt_field(&mut d, "suffix", suffix);
      }
      Kind::Quoted {
        content,
        delims,
        prefix,
        suffix,
      } => {
        d.field("content", content);
        req_field(&mut d, "open", &delims.0);
        req_field(&mut d, "close", &delims.1);
        opt_field(&mut d, "prefix", prefix);
        opt_field(&mut d, "suffix", suffix);
      }
      Kind::Digital { mant, exps, suffix } => {
        d.field("mant", mant).field("exps", exps);
        opt_field(&mut d, "suffix", suffix);
      }
      Kind::Delimited { delims, tokens } => {
        req_field(&mut d, "open", &delims.0);
        req_field(&mut d, "close", &delims.1);
        d.field("tokens", tokens);
      }
      _ => {}
    };

    req_field(&mut d, "span", &self.span);
    vec_field(&mut d, "comments", &self.comments);
    d.finish()
  }
}

struct MatchState<'a> {
  ctx: &'a Context,
  errors: String,
  stack: Vec<usize>,
  error_count: usize,
}

impl MatchState<'_> {
  fn error(&mut self, msg: impl Display) {
    use std::fmt::Write;

    self.error_count += 1;
    if self.error_count > 10 {
      return;
    }

    self.errors.push_str("at stream");
    for i in &self.stack {
      let _ = write!(self.errors, "[{}]", i);
    }
    let _ = writeln!(self.errors, ": {msg}");
  }

  fn match_spans(&mut self, what: &str, text: &Text, span: Span) {
    if !text.recognizes(span, self.ctx) {
      self.error(f!("wrong {what}; want {:?}, got {:?}", text, span));
    }
  }

  fn match_options(
    &mut self,
    what: &str,
    text: Option<&Text>,
    span: Option<Span>,
  ) {
    if text.is_none() && span.is_none() {
      return;
    }

    if !text
      .zip(span)
      .is_some_and(|(t, s)| t.recognizes(s, self.ctx))
    {
      self.error(f!("wrong {what}; want {:?}, got {:?}", text, span));
    }
  }

  fn match_any_options<T: fmt::Debug, U: fmt::Debug>(
    &mut self,
    what: &str,
    text: Option<T>,
    span: Option<U>,
    eq: impl FnOnce(&T, &U) -> bool,
  ) {
    if text.is_none() && span.is_none() {
      return;
    }

    if !text
      .as_ref()
      .zip(span.as_ref())
      .is_some_and(|(t, s)| eq(t, s))
    {
      self.error(f!("wrong {what}; want {:?}, got {:?}", text, span));
    }
  }

  fn finish(mut self) -> Result<(), String> {
    use std::fmt::Write;

    if self.error_count > 10 {
      let _ =
        writeln!(self.errors, "... and {} more errors", self.error_count - 1);
    }

    if self.error_count > 0 {
      return Err(self.errors);
    }
    Ok(())
  }
}

fn zip_eq<Ours: IntoIterator, Theirs: IntoIterator>(
  state: &mut MatchState,
  ours: Ours,
  theirs: Theirs,
  mut cb: impl FnMut(&mut MatchState, Ours::Item, Theirs::Item),
) {
  let mut ours = ours.into_iter();
  let mut theirs = theirs.into_iter();
  state.stack.push(0);
  loop {
    let ours = ours.next();
    let theirs = theirs.next();
    if ours.is_none() && theirs.is_none() {
      state.stack.pop();
      break;
    }

    if let (Some(ours), Some(theirs)) = (ours, theirs) {
      cb(state, ours, theirs);

      *state.stack.last_mut().unwrap() += 1;
      continue;
    }

    state.error("unequal lengths");
    state.stack.pop();
    break;
  }
}
