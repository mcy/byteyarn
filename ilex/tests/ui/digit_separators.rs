use ilex::rule::*;
use ilex::testing;
use ilex::Context;

#[test]
fn digit_separators() {
  let ctx = Context::new();
  let report = ctx.new_report();
  let _ = ctx
    .new_file(
      "<input>",
      "
all_ok@_123_._456_e_789_._012_
no_prefix@_123_._456_e_789_._012_
no_suffix@_123_._456_e_789_._012_
no_point@_123_._456_e_789_._012_
no_exp@_123_._456_e_789_._012_
    ",
    )
    .lex(Spec::get().spec(), &report);

  testing::check_report(&report, "tests/ui/goldens/digit_separators.stdout");
}

ilex::spec! {
  struct Spec {
    n0: Digital = Digital::new(10)
      .prefix("all_ok@")
      .point_limit(0..3)
      .exponent("e", Digits::new(10).point_limit(0..3))
      .separator_with("_",
        SeparatorCornerCases {
          prefix: true,
          suffix: true,
          around_point: true,
          around_exp: true,
        }),
    n1: Digital = Digital::new(10)
    .prefix("no_prefix@")
    .point_limit(0..3)
    .exponent("e", Digits::new(10).point_limit(0..3))
    .separator_with("_",
      SeparatorCornerCases {
        prefix: false,
        suffix: true,
        around_point: true,
        around_exp: true,
      }),
    n2: Digital = Digital::new(10)
    .prefix("no_suffix@")
    .point_limit(0..3)
    .exponent("e", Digits::new(10).point_limit(0..3))
    .separator_with("_",
      SeparatorCornerCases {
        prefix: true,
        suffix: false,
        around_point: true,
        around_exp: true,
      }),
    n3: Digital = Digital::new(10)
    .prefix("no_point@")
    .point_limit(0..3)
    .exponent("e", Digits::new(10).point_limit(0..3))
    .separator_with("_",
      SeparatorCornerCases {
        prefix: true,
        suffix: true,
        around_point: false,
        around_exp: true,
      }),
    n4: Digital = Digital::new(10)
    .prefix("no_exp@")
    .point_limit(0..3)
    .exponent("e", Digits::new(10).point_limit(0..3))
    .separator_with("_",
      SeparatorCornerCases {
        prefix: true,
        suffix: true,
        around_point: true,
        around_exp: false,
      }),
  }
}
