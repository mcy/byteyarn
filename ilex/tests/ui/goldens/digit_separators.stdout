error: unexpected digit separator in `no_prefix@`-prefixed number
 --> <input>:3:11
  |
3 | no_prefix@_123_._456_e_789_._012_
  |           ^
  |

error: unexpected digit separator in `no_suffix@`-prefixed number
 --> <input>:4:33
  |
4 | no_suffix@_123_._456_e_789_._012_
  |                                 ^
  |

error: unexpected digit separator in `no_point@`-prefixed number
 --> <input>:5:15
  |
5 | no_point@_123_._456_e_789_._012_
  |               ^
  |

error: unexpected digit separator in `no_point@`-prefixed number
 --> <input>:5:16
  |
5 | no_point@_123_._456_e_789_._012_
  |                ^
  |

error: unexpected digit separator in `no_point@`-prefixed number
 --> <input>:5:27
  |
5 | no_point@_123_._456_e_789_._012_
  |                           ^
  |

error: unexpected digit separator in `no_point@`-prefixed number
 --> <input>:5:28
  |
5 | no_point@_123_._456_e_789_._012_
  |                            ^
  |

error: unexpected digit separator in `no_exp@`-prefixed number
 --> <input>:6:19
  |
6 | no_exp@_123_._456_e_789_._012_
  |                   ^
  |

error: unexpected digit separator in `no_exp@`-prefixed number
 --> <input>:6:20
  |
6 | no_exp@_123_._456_e_789_._012_
  |                    ^
  |

error: aborting due to 8 errors
